core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java

```diff
@@ -17,87 +17,150 @@
+      @NonNull Iterable<Entry<Node, List<Throwable>>> errors) {
     super(message, executionInfo, null, true);
-    this.errors = errors;
+    this.errors = toDeepImmutableMap(errors);
   }
 
-  private AllNodesFailedException(Map<Node, Throwable> errors) {
+  private AllNodesFailedException(Collection<Entry<Node, List<Throwable>>> errors) {
     this(
         buildMessage(
             String.format("All %d node(s) tried for the query failed", errors.size()), errors),
         null,
         errors);
   }
 
-  private static String buildMessage(String baseMessage, Map<Node, Throwable> errors) {
+  private static String buildMessage(
+      String baseMessage, Collection<Entry<Node, List<Throwable>>> errors) {
     int limit = Math.min(errors.size(), 3);
-    String details =
-        Joiner.on(", ").withKeyValueSeparator(": ").join(Iterables.limit(errors.entrySet(), limit));
-
+    Iterator<Entry<Node, List<Throwable>>> iterator = errors.iterator();
+    StringBuilder details = new StringBuilder();
+    while (iterator.hasNext()) {
+      Entry<Node, List<Throwable>> entry = iterator.next();
+      details.append(entry.getKey()).append(": ").append(entry.getValue());
+      if (iterator.hasNext()) {
+        details.append(", ");
+      }
+    }
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        "%s (showing first %d, use getAllErrors() for more): %s", baseMessage, limit, details);
   }
 
-  /** The details of the individual error on each node. */
+  /**
+   * An immutable map containing the first error on each tried node.
+   *
+   * @deprecated Use {@link #getAllErrors()} instead.
+   */
   @NonNull
+  @Deprecated
   public Map<Node, Throwable> getErrors() {
+    ImmutableMap.Builder<Node, Throwable> builder = ImmutableMap.builder();
+    for (Node node : errors.keySet()) {
+      List<Throwable> nodeErrors = errors.get(node);
+      if (!nodeErrors.isEmpty()) {
+        builder.put(node, nodeErrors.get(0));
+      }
+    }
+    return builder.build();
+  }
+
+  /** An immutable map containing all errors on each tried node. */
+  @NonNull
+  public Map<Node, List<Throwable>> getAllErrors() {
     return errors;
   }
 
   @NonNull
   @Override
   public DriverException copy() {
-    return new AllNodesFailedException(getMessage(), getExecutionInfo(), errors);
+    return new AllNodesFailedException(getMessage(), getExecutionInfo(), errors.entrySet());
   }
 
   @NonNull
   public AllNodesFailedException reword(String newMessage) {
     return new AllNodesFailedException(
-        buildMessage(newMessage, errors), getExecutionInfo(), errors);
+        buildMessage(newMessage, errors.entrySet()), getExecutionInfo(), errors.entrySet());
+  }
+
+  private static Set<Entry<Node, List<Throwable>>> groupByNode(
+      Iterable<Entry<Node, Throwable>> errors) {
+    // no need for immutable collections here
+    Map<Node, List<Throwable>> map = new LinkedHashMap<>();
```

Comment:
Isn't it better to just use guaga `MultiMap`?

Reply:
Not possible, we cannot expose Guava in the API (it's shaded).